contracts/btc-light-client/src/bitcoin.rs

```diff
@@ -1,54 +1,277 @@
+
     for (i, new_header) in new_headers.iter().enumerate() {
-        let last_btc_header: BlockHeader = deserialize(last_header.header.as_ref())?;
-        let btc_header: BlockHeader = deserialize(new_header.header.as_ref())?;
+        let prev_block_header: BlockHeader = deserialize(last_header.header.as_ref())?;
+        let block_header: BlockHeader = deserialize(new_header.header.as_ref())?;
+
+        // Check whether the headers form a chain.
+        if block_header.prev_blockhash != prev_block_header.block_hash() {
+            return Err(HeaderError::PrevHashMismatch {
+                got: block_header.prev_blockhash,
+                expected: prev_block_header.block_hash(),
+            });
+        }
 
-        // validate whether btc_header extends last_btc_header
-        babylon_bitcoin::pow::verify_next_header_pow(chain_params, &last_btc_header, &btc_header)?;
+        check_header(
+            storage,
+            chain_params,
+            last_header.height,
+            &prev_block_header,
+            &block_header,
+        )?;
 
-        let header_work = btc_header.work();
         let cum_work = total_work(new_header.work.as_ref())?;
 
         // Validate cumulative work
-        if cum_work_old + header_work != cum_work {
-            return Err(ContractError::BTCWrongCumulativeWork(
-                i,
-                cum_work_old + header_work,
-                cum_work,
-            ));
+        let cum_work_new = cum_work_old + block_header.work();
+        if cum_work_new != cum_work {
+            return Err(HeaderError::WrongCumulativeWork(i, cum_work_new, cum_work));
         }
-        cum_work_old = cum_work;
+
         // Validate height
         if new_header.height != last_header.height + 1 {
-            return Err(ContractError::BTCWrongHeight(
+            return Err(HeaderError::WrongHeight(
                 i,
                 last_header.height + 1,
                 new_header.height,
             ));
         }
 
         // this header is good, verify the next one
+        cum_work_old = cum_work;
         last_header = new_header.clone();
     }
     Ok(())
 }
 
+/// This functions mirrors the `checkHeader` in babylon golang implmentation.
+///
+/// https://github.com/babylonlabs-io/babylon/blob/48617fb852e9cae4ea7ea38c80793cdcb6f2668c/x/btclightclient/types/btc_light_client.go#L416
+fn check_header(
+    storage: &dyn Storage,
+    chain_params: &Params,
+    prev_block_height: u32,
+    prev_block_header: &BlockHeader,
+    header: &BlockHeader,
+) -> Result<(), HeaderError> {
+    check_block_header_context(
+        storage,
+        chain_params,
+        prev_block_height,
+        prev_block_header,
+        header,
+    )?;
+
+    // Perform proof-of-work check as specified in:
+    // https://pkg.go.dev/github.com/btcsuite/btcd@v0.24.2/blockchain#CheckBlockHeaderSanity
+    check_proof_of_work(chain_params, header)?;
```

Comment:
It seems the sanity check includes both pow check and timestamp check. Why do we skip timestamp check here?

Reply:
See this comment https://github.com/babylonlabs-io/babylon/blob/3e474546ab80b2696a66c10506f84bb17ce53bca/types/btcutils.go#L17-L21